core/modules/replica/ControlThread.h

```diff
@@ -0,0 +1,368 @@
+     * to the method.
+     *
+     * @param jobName
+     *   the name of a job
+     *
+     * @param Fargs
+     *   job-specific variadic parameters
+     * 
+     * @throws
+     *   ControlThreadStopped when the thread cancellation request was detected
+     */
+    template <class T, typename...Targs>
+    void launch(std::string const& jobName,
+                Targs... Fargs) {
+
+        info(jobName);
+
+        // Launch the jobs
+
+        std::vector<Job::Ptr> jobs;
+        std::atomic<size_t> numFinishedJobs{0};
+
+        std::string const parentJobId;  // no parent for these jobs
+
+        for (auto&& family: serviceProvider()->config()->databaseFamilies()) {
+            auto job = T::create(
+                family,
+                Fargs...,
+                controller(),
+                parentJobId,
+                [&numFinishedJobs](typename T::Ptr const& job) {
+                    ++numFinishedJobs;
+                    // FIXME: analyze job status and report it here
+                }
+            );
+            job->start();
+            jobs.push_back(job);
+        }
+
+        // Track the completion of all jobs
+
+        track<Job>(jobName,
+                   jobs,
+                   numFinishedJobs);
+    }
+
+   /**
+     * Launch Qserv synchronization jobs.
+     *
+     * @param qservSyncTimeoutSec
+     *   the number of seconds to wait before a completion of the synchronization
+     *   operation.
+     * 
+     * @param forceQservSync
+     *   (optional) force Qserv synchronization if 'true'
+     * 
+     * @throws
+     *   ControlThreadStopped when the thread cancellation request was detected
+     *
+     * @see QservSyncJob
+     */
+    void sync(unsigned int qservSyncTimeoutSec,
+              bool forceQservSync=false);
+
+    /**
+     * Track the completion of all jobs. Also monitor the thread cancellation
+     * condition while tracking the jobs. When such condition will be seen
+     * all jobs will be canceled. The tracking will be done with an interval
+     * of 1 second.
+     *
+     * @param jobName
+     *   the name of a job
+     *
+     * @param jobs
+     *   the collection of jobs to be tracked
+     *
+     * @param numFinishedJobs
+     *   the counter of completed jobs
+     */
+    template <class T>
+    void track(std::string const& jobName,
+               std::vector<typename T::Ptr> const& jobs,
+               std::atomic<size_t> const& numFinishedJobs) {
+
+        info(jobName + ": tracking started");
+    
+        util::BlockPost blockPost(1000, 1001);  // ~1 second wait time between iterations
+    
+        while (numFinishedJobs != jobs.size()) {
+            if (stopRequested()) {
+                for (auto&& job: jobs) {
+                    job->cancel();
+                }
+                info(jobName + ": tracking aborted");
+                throw ControlThreadStopped();
+            }
+            blockPost.wait();
```

Comment:
I think it would be better to use the asio service to manage things that these thread controller classes are doing. Can we spend some time on it (with Fritz, who knows asio better than I do) during the meeting on Wednesday?

Reply:
Surely, we can discuss it! I guess, your question is about the threading oder of the Master Controller application. It's rather trivial:
- there is a (configurable) number of the BOOST ASIO threads which are used for asynchronous operations (completion of the network I/O requests, timers, scheduling delayed activities)
- there 1 dedicated thread launched by `ControlThread::start`  for each of the following subclasses of `ControlThread`: `HttpThread`, 'ReplicationThread`, 'HealthMonitorThread` and `DeleteWorkerThread`.
- and there is obviously the `main` thread

A pool of the BOOST ASIO threads is configured based on a scale of the Replication System's setup, hardware capabilities (the number of cores, memory) of a machine where the Master Controller runs.

The `ControlThread`s threads are managed differently. They need to be independent of the BOOST ASIO's threads. Besides, they're launched/stopped dynamically depending on a status of the Replication System. Note that BOOST ASIO's threads are always there. And those (the later) may be all taken by some on-going processing.

Does it makes things more clear to you? Ot it makes things even more confusing?  :-)